core/src/smartcontracts/isi/permissions.rs

```diff
@@ -368,52 +699,209 @@ fn check_query_in_instruction<W: WorldTrait>(
     }
 }
 
-impl<W: WorldTrait> From<CheckNested<W, Instruction>> for IsAllowedBoxed<W, Instruction> {
-    fn from(validator: CheckNested<W, Instruction>) -> Self {
-        Box::new(validator)
+fn check_all_validators_have_the_same_type(validators: &[IsAllowedBoxed]) -> Result<()> {
+    let first_type = if let Some(first) = validators.first() {
+        first.validator_type()
+    } else {
+        return Ok(());
+    };
+
+    for validator in validators.iter().skip(1) {
+        let validator_type = validator.validator_type();
+        if validator_type != first_type {
+            return Err(ValidatorTypeMismatch::expected(first_type)
+                .found(validator_type)
+                .into());
+        }
     }
+
+    Ok(())
 }
 
 /// A container for multiple permissions validators. It will succeed if all validators succeed.
-pub struct AllShouldSucceed<W: WorldTrait, O: NeedsPermission> {
-    validators: Vec<IsAllowedBoxed<W, O>>,
+#[derive(Debug, Clone, Serialize)]
+pub struct AllShouldSucceed {
+    validators: Vec<IsAllowedBoxed>,
 }
 
-impl<W: WorldTrait, O: NeedsPermission> IsAllowed<W, O> for AllShouldSucceed<W, O> {
+impl AllShouldSucceed {
+    /// Creates new [`AllShouldSucceed`]
+    ///
+    /// # Errors
+    /// If provided validators have different types.
+    /// Type of the first element in the `validators` is considered exemplary
+    pub fn new(validators: Vec<IsAllowedBoxed>) -> Result<Self> {
+        check_all_validators_have_the_same_type(&validators)?;
+        Ok(Self { validators })
+    }
+
+    fn check_type(&self, validator_type: ValidatorType) -> Result<()> {
+        if let Ok(self_type) = self.validator_type() {
+            if self_type != validator_type {
+                return Err(ValidatorTypeMismatch::expected(validator_type)
```

Comment:
The reason why I would prefer the usage of the generic is also because of the syntax: 
Compare: 
```rust
Mismatch<ValidatorType> {
   expected: validator_type,
   actual: self_type
}

```

Reply:
Yeah, I was thinking about accidental parameter swapping and completely forgot about this construction syntax